### PR TITLE
Add support for 8da4w quantization

### PR DESCRIPTION
**Pull request by andrewor14**
Summary: Add a new quantization for users to quantize their models using int8 per token dynamic activation + int4 per axis grouped weight quantization.

Test Plan:
```
tune run quantize --config quantization \
    quantizer._component_=torchtune.utils.quantization.Int8DynActInt4WeightQuantizer \
    quantizer.groupsize=256

tune run eleuther_eval --config eleuther_eval \
    checkpointer._component_=torchtune.utils.FullModelTorchTuneCheckpointer \
    checkpointer.checkpoint_files=[hf_model_0001_2-8da4w.pt] \
    quantizer._component_=torchtune.utils.quantization.Int8DynActInt4WeightQuantizer \
    quantizer.groupsize=256
```

Quantize output:
```
2024-04-26:13:09:20,852 INFO     [quantize.py:98] Time for quantization: 7.49 sec
2024-04-26:13:09:20,852 INFO     [quantize.py:99] Memory used: 15.30 GB
2024-04-26:13:09:27,625 INFO     [quantize.py:112] Model checkpoint of size 7.08 GB saved to /home/andrewor/logs/tune/saved-4-25/full_1713990537_complete/quantize_output/hf_model_0001_2-8da4w.pt
```

Eval output:
```
100%|██████████| 5882/5882 [17:05<00:00,  5.74it/s]
2024-04-26:13:50:47,214 INFO     [eleuther_eval.py:196] Eval completed in 1040.88 seconds.
2024-04-26:13:50:47,214 INFO     [eleuther_eval.py:198] truthfulqa_mc2: {'acc,none': 0.4578906989909618, 'acc_stderr,none': 0.01542454257740726, 'alias': 'truthfulqa_mc2'}
```

Reviewers: jerryzh168, kartikayk, ebsmothers

Subscribers: jerryzh168, kartikayk, ebsmothers, supriyar

### CODE REVIEW

**Review by ebsmothers — APPROVED** [review]

File: torchtune/utils/quantization.py [hunk 1]
```diff
@@ -11,13 +11,15 @@
...
     Int4WeightOnlyQuantizer,
+    Int8DynActInt4WeightQuantizer,
     Quantizer,
 )
 
 __all__ = [
     "Int4WeightOnlyQuantizer",
     "Int4WeightOnlyGPTQQuantizer",
     "Int8WeightOnlyQuantizer",
+    "Int8DynActInt4WeightQuantizer",
```
Comment (rohan-varma) [inline torchtune/utils/quantization.py:22]: What are the benefits of this int8 dynamic act. quantizer over existing quantizers in torchtune? Do we plan to offer tutorials / additional documentation clarifying to users how to think about which quantizers to use?

Comment (andrewor14) [reply torchtune/utils/quantization.py:22]: This is the only form of quantization supported by QAT right now, so this PR is adding the same form of quantization for PTQ for comparison. As for the documentation of which quantizer to use, I think we can add it in the future. Right now I'm not sure if we have any even in torchao or gpt-fast comparing 8da4w to weight only quantization. cc @jerryzh168 @HDCharles

**Review by rohan-varma — APPROVED** [review]
Sounds good! Please feel free to land.

**Comment by andrewor14** [comment]
@rohan-varma @ebsmothers can you help me land? Looks like I don't have write access

**Comment by ebsmothers** [comment]
Done, thanks for adding this!

**Pull request merged**